src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,70 @@ data CodingFailureMode
 replacementChar :: Char
 replacementChar = '\xFFFD'
 
+data UTF8CharDecodeState a
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+-- XXX We can chain this with another parser to propagate the error and act
+-- accordingly.
+{-# INLINE parseCharUtf8D' #-}
+parseCharUtf8D' :: Monad m => ParserD.Parser m Word8 Char
+parseCharUtf8D' =
```

Comment:
We can make the parser accept CodingFailure as an argument and do error handling appropriately. The higher level parser wrapping this routine can choose an appropriate value for that argument.

Reply:
I find this to be more elegant.
```
parseCharUtfD = parseCharUtfD' <|> fromPure replacementChar
```

Reply:
All I want is to reuse the same implementation for fold. The low level routine `parserCharInternal options -> Parser m Word8 Char` can be used to express both.

```
parserCharUtf8 = parserCharInternal ErrorOnCodingFailure
foldCharUtf8 = toFold $ parserCharInternal DropOnCodingFailure
```

Now you can use parseCharUtf8 in an alternative composition or whatever. Not that parserCharUtf8 would use the Error constructor in all the error cases and not "error" so that it is always a proper parser, the toFold routine then converts the Error to error.